### PR TITLE
Multiple lists support in Mac

### LINKED ISSUE #125: Support multiple lists in Mac app
Multiple list support was added in the iOS version in #102. In that PR, I added enough "support" to the Mac version for it to not break, but the Mac app's UI still just shows the initial list.

@stel I don't think this is super pressing, since the Mac app continues to work, but it'd be nice to do eventually.

### PR DESCRIPTION
**Pull request by stel**
The UI still needs to be polished the code should also be cleaned up a bit, so will keep at as WIP. I'll be very happy to hear any feedback even on that stage.

Fixes #125.

/cc @TimOliver, @jpsim

### CODE REVIEW

**Comment by stel** [comment]
Well, seems like it's ready so I'd really happy to hear your feedback!

Some notes here:

1. To rename list hold mouse pointer on it's cell for ~1 sec until cursor is changed to a text cursor (like in Clear app) then click to edit.
2. `Cmd+N` could also be used to create new Task/List. More shortcuts coming soon.
3. I don't really like the title in toolbar, it should be improved later
4. `ItemCellView.swift` could be compared with `TaskCellView.swift` from `master` to track changes.

/cc @jpsim, @TimOliver

**Comment by jpsim** [comment]
Could you please rebase or merge `master` to make it easier to review this? Thanks.

File: RealmTasks Apple/RealmTasks macOS/NSView+Animations.swift [hunk 1]
```diff
@@ -20,13 +20,48 @@ import Cocoa
...
+    // Convenient method for trailing closure syntax without completion handler
+    static func animate(duration duration: NSTimeInterval = defaultAnimationDuration, timingFunction: CAMediaTimingFunction = defaultAnimationTimingFunction, animations: () -> Void) {
+        animate(duration: duration, timingFunction: timingFunction, animations: animations, completion: nil)
+    }
+
+}
+
+extension CAMediaTimingFunction {
+
+    static func lineral() -> CAMediaTimingFunction {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/NSView+Animations.swift]: should be `linear()`

File: RealmTasks Apple/RealmTasks macOS/NSView+Animations.swift [hunk 2]
```diff
@@ -20,13 +20,48 @@ import Cocoa
...
             context.allowsImplicitAnimation = true
 
+            context.duration = duration
+            context.timingFunction = timingFunction
+
             animations()
         }, completionHandler: completion)
     }
 
+    // Convenient method for trailing closure syntax without completion handler
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/NSView+Animations.swift]: why is this necessary given that the other one has a `completion` default value of `nil`?

Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/NSView+Animations.swift]: nevermind

**Comment by stel** [comment]
Done, thanks @jpsim!

**Comment by TimOliver** [comment]
Thanks for the PR @stel! I had a look at it/play with it, and it looks great!

However, I've discovered a reproducible crash in it. In either the tasks, or the lists view, if you try and re-order the items, it's pretty easy to trigger this fatal error:

screen shot 2016-10-11 at 2 14 46 pm

**Comment by stel** [comment]
Ooops, thanks, @TimOliver!

What OS version you use? I wasn't able to reproduce that crash but made some extra check to potentially prevent it, could you please check it?

**Comment by TimOliver** [comment]
Thanks @stel!

 I was running OS X 10.11 still, but I updated to macOS 10.12 last night. I tried it again just now with your changes and it all runs perfectly! :)

**Comment by stel** [comment]
Cool, thanks @TimOliver!

@jpsim any other things to fix here? what do you think about merging it?

**Comment by jpsim** [comment]
With this branch, registering a user I get the following uncaught exception:

```
realm::RealmFileException: Unable to open a realm at path '/Users/jp/Library/Application Support/io.realm.realmtasks.macos/realm-object-server/io.realm.object-server-metadata/sync_metadata.realm.management': make_dir() failed: No such file or directory.
```

**Comment by jpsim** [comment]
List title doesn't appear to be centered:

image

**Comment by jpsim** [comment]
The "lists" and "add" buttons could also afford to be smaller. They look to big in my opinion.

File: RealmTasks Apple/RealmTasks macOS/ContainerViewController.swift [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        guard let toolbar = view.window?.toolbar else {
+            return
+        }
+
+        if let titleLabel = toolbar.itemWithIdentifier("TitleLabel")?.view as? NSTextField {
+            titleLabel.stringValue = (list as? CellPresentable)?.text ?? "Lists"
+        }
+
+        if list is CellPresentable {
+            if !toolbar.hasItemWithIdentifier("ShowAllListsButton") {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ContainerViewController.swift]: `"ShowAllListsButton"` is hardcoded 3 times here, could extract it into a constant just above.

File: RealmTasks Apple/RealmTasks macOS/ItemCellView.swift [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+// FIXME: This file should be split up.
+// swiftlint:disable file_length
+
+import Cocoa
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ItemCellView.swift]: we've sorted imports elsewhere in the project, could do that here too.

File: RealmTasks Apple/RealmTasks macOS/ItemCellView.swift [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+        deleteIconView.frame.size.width = iconWidth
+        deleteIconView.frame.origin.x = bounds.width - deleteIconView.bounds.width - iconOffset
+        deleteIconView.autoresizingMask = [.ViewMinXMargin, .ViewHeightSizable]
+        addSubview(deleteIconView, positioned: .Below, relativeTo: contentView)
+    }
+
+    private func setupContentView() {
+        addSubview(contentView)
+
+        contentView.frame = bounds
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ItemCellView.swift]: why isn't this using autolayout?

File: RealmTasks Apple/RealmTasks macOS/ItemCellView.swift [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+    }
+
+}
+
+// MARK: NSGestureRecognizerDelegate
+
+extension ItemCellView: NSGestureRecognizerDelegate {
+
+    func gestureRecognizerShouldBegin(gestureRecognizer: NSGestureRecognizer) -> Bool {
+        let currentlyEditingTextField = ((window?.firstResponder as? NSText)?.delegate as? NSTextField)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ItemCellView.swift]: extra parens

File: RealmTasks Apple/RealmTasks macOS/ItemCellView.swift [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+        invalidateIntrinsicContentSize()
+    }
+
+}
+
+// MARK: Private Extensions
+
+private extension NSTextField {
+
+    func strike(fraction: Double = 1) {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ItemCellView.swift]: These functions could be mostly shared across iOS and macOS if they were on `NSAttributedString` instead of `NSTextField`/`CellTextView`.

File: RealmTasks Apple/RealmTasks macOS/ListCellView.swift [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+    private func setupTextView() {
+        constrain(textView, badgeView, replace: textViewConstraintGroup) { textView, badgeView in
+            textView.left == textView.superview!.left + 13
+            textView.top == textView.superview!.top + 11
+            textView.bottom == textView.superview!.bottom - 11
+            textView.right == badgeView.left - 13
+        }
+
+        // It's not a good idea but seems to work
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ListCellView.swift]: 😂

File: RealmTasks Apple/RealmTasks macOS/ListViewController.swift [hunk 9]
```diff
@@ -103,42 +136,52 @@ class TaskListViewController: NSViewController {
...
+        endEditingCells()
 
-    @IBAction func newTask(sender: AnyObject?) {
-        // Commit any currently editing cells
-        currentlyEditingCellView?.window?.makeFirstResponder(nil)
-
-        try! items.realm?.write {
-            items.insert(Task(), atIndex: 0)
+        try! list.realm?.write {
+            self.list.items.insert(ItemType(), atIndex: 0)
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ListViewController.swift]: unnecessary `self.`

File: RealmTasks Apple/RealmTasks macOS/ListViewController.swift [hunk 10]
```diff
@@ -103,42 +136,52 @@ class TaskListViewController: NSViewController {
...
         }
 
-        NSView.animateWithDuration(0.2, animations: {
+        self.animating = true
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ListViewController.swift]: unnecessary `self.`

File: RealmTasks Apple/RealmTasks macOS/ListViewController.swift [hunk 11]
```diff
@@ -103,42 +136,52 @@ class TaskListViewController: NSViewController {
...
+        return validateSelector(toolbarItem.action)
+    }
+
+    override func validateMenuItem(menuItem: NSMenuItem) -> Bool {
+        return validateSelector(menuItem.action)
     }
-}
 
-// MARK: Reordering
+    private func validateSelector(selector: Selector) -> Bool {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ListViewController.swift]: nice

File: RealmTasks Apple/RealmTasks macOS/ListViewController.swift [hunk 12]
```diff
@@ -151,19 +194,29 @@ extension TaskListViewController {
             return
         }
 
+        tableView.enumerateAvailableRowViewsUsingBlock { _, row in
+            if let view = self.tableView.viewAtColumn(0, row: row, makeIfNecessary: false) as? ItemCellView {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ListViewController.swift]: unnecessary `self.`

File: RealmTasks Apple/RealmTasks macOS/ListViewController.swift [hunk 13]
```diff
@@ -346,177 +489,152 @@ extension TaskListViewController: NSTableViewDelegate {
...
+    // MARK: ItemCellViewDelegate
 
-    func cellView(view: TaskCellView, didComplete complete: Bool) {
-        guard let (item, index) = findItemForCellView(view) else {
+    func cellView(view: ItemCellView, didComplete complete: Bool) {
+        guard let (tmpItem, index) = findItemForCellView(view) else {
             return
         }
 
+        // FIXME: workaround for tuple mutability
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/ListViewController.swift]: Not sure what there is to fix here. Kind of how Swift works. Could do this if you really want:

```swift
guard let itemAndIndex = findItemForCellView(view) else {
    return
}
var item = itemAndIndex.0
let index = itemAndIndex.1
```

File: RealmTasks Apple/RealmTasks macOS/NSView+TrackingAreas.swift [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+        let trackingArea = NSTrackingArea(rect: rect, options: options, owner: self, userInfo: nil)
+
+        addTrackingArea(trackingArea)
+
+        return trackingArea
+    }
+
+    public func resetTrackingAreas() {
+        for trackingArea in trackingAreas {
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/NSView+TrackingAreas.swift]: Up to you, but `trackingAreas.forEach(removeTrackingArea)` works here too.

File: RealmTasks Apple/RealmTasks macOS/ListViewController.swift (hunk 13 above)
Comment (stel) [inline RealmTasks Apple/RealmTasks macOS/ListViewController.swift]: Nice, thanks!

File: RealmTasks Apple/RealmTasks macOS/ItemCellView.swift (hunk 5 above)
Comment (stel) [inline RealmTasks Apple/RealmTasks macOS/ItemCellView.swift]: It's just more handy to change `contentVIew` frame directly than using `LayoutGroups` while handling swipe because there are also 2 views with `done` and `remove` icons.

**Comment by stel** [comment]
```
realm::RealmFileException: Unable to open a realm at path '/Users/jp/Library/Application Support/io.realm.realmtasks.macos/realm-object-server/io.realm.object-server-metadata/sync_metadata.realm.management': make_dir() failed: No such file or directory.
```

This is because this file is deleted when server starts (see #302), should be fixed in a separate PR.

**Comment by jpsim** [comment]
Title bar flickers when being resized now, and title is often "fuzzy", probably when it's being centered in half-pixels. This crappy gif may not convey exactly the impact this has, but you can likely try yourself:

flicker_titlebar

This only happens on non-retina. Seems to render perfectly on retina resolutions AFAICT.

I'm ok if you want to merge this as-is and want to track that as a separate issue.

**Comment by stel** [comment]
Thanks a lot @jpsim and @TimOliver! I've fixed the rendering issue for non-retina as well so merging 🚀

File: RealmTasks Apple/RealmTasks macOS/TitleView.swift [hunk 15]
```diff
@@ -66,6 +66,9 @@ class TitleView: NSView {
         // ... and vertically in self
         frame.origin.y = (bounds.height - textField.frame.height) / 2
 
+        // Float values leads to blurred drawing on non-retina screens
+        frame.makeIntegralInPlace()
```
Comment (jpsim) [inline RealmTasks Apple/RealmTasks macOS/TitleView.swift]: I didn't know this existed! Neat 👌
